File: RScan/Python/scan/scan.py
import cv2 as cv
import numpy as np

'''
 Function to map values in range [in_min, in_max] to the range [out_min, out_max]
'''
def map(x, in_min, in_max, out_min, out_max):
	return (x - in_min) * (out_max - out_min) / (in_max - in_min) + out_min


# global variables for image ,blackPoint, whitePoint
img = None; whitePoint = None; blackPoint = None;


'''
 * High Pass Filter
 * Output of HPF depends on the kernel size provided as input argument
 * Links to the docuementation:
 *  Introduction: https://github.com/sourabhkhemka/DocumentScanner/wiki/Scan:-Introduction
 *  HPF: https://github.com/sourabhkhemka/DocumentScanner/wiki/GCMODE
 *
'''
def highPassFilter(kSize):
	global img
	
	print("applying high pass filter")
	
	if not kSize%2:
		kSize +=1
		
	kernel = np.ones((kSize,kSize),np.float32)/(kSize*kSize)
	
	filtered = cv.filter2D(img,-1,kernel)
	
	filtered = img.astype('float32') - filtered.astype('float32')
	filtered = filtered + 127*np.ones(img.shape, np.uint8)
	
	filtered = filtered.astype('uint8')
	
	img = filtered
	
	# "img" now contains high pass filtered image.

	
	
'''
 * Method to select black point in the image
 *
 * Links to documentation: 
 *  Introduction: https://github.com/sourabhkhemka/DocumentScanner/wiki/Scan:-Introduction
 *  Black Point Select: https://github.com/sourabhkhemka/DocumentScanner/wiki/Black-Point-Select
''' 
def blackPointSelect():
	global img
	
	print("adjusting black point for final output ...")
	
	# refer repository's wiki page for detailed explanation

	img = img.astype('int32')

	img = map(img, blackPoint, 255, 0, 255)

	#if cv.__version__ == '3.4.4':
		#img = img.astype('uint8')

	_, img = cv.threshold(img, 0, 255, cv.THRESH_TOZERO)

	img = img.astype('uint8')
	

'''
 * Method to select whitePoint in the image
 *
 * Links to documentation: 
 *  Introduction: https://github.com/sourabhkhemka/DocumentScanner/wiki/Scan:-Introduction
 *  White Point Select: https://github.com/sourabhkhemka/DocumentScanner/wiki/White-Point-Select
'''
def whitePointSelect():
	global img
	
	print("white point selection running ...")

	# refer repository's wiki page for detailed explanation

	_,img = cv.threshold(img, whitePoint, 255, cv.THRESH_TRUNC)

	img = img.astype('int32')
	img = map(img, 0, whitePoint, 0, 255)
	img = img.astype('uint8')
	
	
'''
 * Method to select black point in the image
 *
 * Links to documentation: 
 *  Introduction: https://github.com/sourabhkhemka/DocumentScanner/wiki/Scan:-Introduction
 *  Black Point Select: https://github.com/sourabhkhemka/DocumentScanner/wiki/Black-Point-Select
 *
''' 
def blackAndWhite():
	global img
	
	# refer repository's wiki page for detailed explanation

	lab = cv.cvtColor(img, cv.COLOR_BGR2LAB)

	(l,a,b) = cv.split(lab)

	img = cv.add( cv.subtract(l,b), cv.subtract(l,a) )



if __name__ == "__main__":

	img = cv.imread("C:/Users/Sourabh Khemka/Desktop/RScan/ML/dataset/data_img_008.jpg")

	# define values for blackPoint and whitePoint
	blackPoint = 66
	whitePoint = 160
	
	# store desired mode of operation as string
	mode = "GCMODE"

	if mode == "GCMODE":
		highPassFilter(kSize = 51)
		whitePoint = 127
		whitePointSelect()
		blackPointSelect()
	elif mode == "RMODE":
		blackPointSelect()
		whitePointSelect()
	elif mode == "SMODE":
		blackPointSelect()
		whitePointSelect()
		blackAndWhite()



print("\ndone.")

cv.imwrite('C:/Users/Sourabh Khemka/Desktop/RScan/ML/op.jpg', img)

cv.imshow("final", cv.resize(img,None,fx=0.125, fy=0.125, interpolation = cv.INTER_CUBIC))
cv.waitKey(0)

File: RScan/Java/src/scan/Scan.java
package scan;

import java.util.ArrayList;
import java.util.List;

import org.opencv.core.Core;
import org.opencv.core.CvType;
import org.opencv.core.Mat;
import org.opencv.core.Scalar;
import org.opencv.imgproc.Imgproc;

public class Scan {

    // Mat objects to store input image, filtered image from HPF & scanned image
    private Mat inputImg;
    private Mat filtered = new Mat();
    private Mat processedImg = new Mat();

    private int kSize;
    private int blackPoint;
    private int whitePoint;

    // enum type to help select mode of scanning
    public static enum ScanMode{
        GCMODE,
        RMODE,
        SMODE,
    }

    // constructor
    public Scan (Mat image, int kernelSize, int blackPoint, int whitePoint){
        this.inputImg = image.clone();
        this.kSize = kernelSize;
        this.blackPoint = blackPoint;
        this.whitePoint = whitePoint;
    }


    /* High Pass Filter
     * Output of HPF depends on the kernel size provided through the constructor
     * Links to the docuementation:
     *  Introduction: https://github.com/sourabhkhemka/DocumentScanner/wiki/Scan:-Introduction
     *  HPF: https://github.com/sourabhkhemka/DocumentScanner/wiki/GCMODE
     */
    private void highPassFilter() 
    {        
        if ( kSize%2 == 0 )
            kSize++;
        
        Mat kernel = Mat.ones(kSize, kSize, CvType.CV_32FC1);
        kernel = kernel.mul(kernel, 1/((float)kSize * (float)kSize));
        
        Imgproc.filter2D(inputImg, filtered, -1, kernel);

        // Convert both to float to avoid saturation of pixel values to 255
        filtered.convertTo(filtered, CvType.CV_32FC3);
        inputImg.convertTo(inputImg, CvType.CV_32FC3);

        Core.subtract(inputImg, filtered, filtered);

        kernel = Mat.zeros(inputImg.size(), CvType.CV_32FC3);
        kernel.setTo(new Scalar(1,1,1));
        
        Core.multiply(kernel, new Scalar(127.0, 127.0, 127.0), kernel);
        
        Imgproc.cvtColor(filtered,filtered,Imgproc.COLOR_RGBA2RGB);
        
        Core.add(filtered, kernel, filtered);

        filtered.convertTo(filtered, CvType.CV_8UC3);

        // "filtered" now contains high pass filtered image.
    }


    /* Method to select whitePoint in the image
     *
     * Links to documentation: 
     *  Introduction: https://github.com/sourabhkhemka/DocumentScanner/wiki/Scan:-Introduction
     *  White Point Select: https://github.com/sourabhkhemka/DocumentScanner/wiki/White-Point-Select
     */
    private void whitePointSelect(){

        // refer repository's wiki page for detailed explanation

        Imgproc.threshold(processedImg, processedImg, whitePoint, 255, Imgproc.THRESH_TRUNC);

        Core.subtract(processedImg, new Scalar(0, 0, 0), processedImg);
        
        float tmp = (255.0f) / ((float)whitePoint - 0);
        Core.multiply(processedImg, new Scalar(tmp, tmp, tmp), processedImg);
        
    }

    /* Method to select black point in the image
     *
     * Links to documentation: 
     *  Introduction: https://github.com/sourabhkhemka/DocumentScanner/wiki/Scan:-Introduction
     *  Black Point Select: https://github.com/sourabhkhemka/DocumentScanner/wiki/Black-Point-Select
     */    
    private void blackPointSelect(){

        // refer repository's wiki page for detailed explanation

        Core.subtract(processedImg, new Scalar(blackPoint, blackPoint, blackPoint), processedImg);
        
        float tmp = (255.0f) / (255.0f - blackPoint);
        Core.multiply(processedImg, new Scalar(tmp, tmp, tmp), processedImg);
    }

    /* Method to process image in LAB color space to generate black and white images
     *  Wiki link: https://github.com/sourabhkhemka/DocumentScanner/wiki/SMODE:-Black-&-White
     */
    private void blackAndWhite(){

        // refer repository's wiki page for detailed explanation
        List <Mat> lab = new ArrayList<>();
        Mat subA = new Mat();
        Mat subB = new Mat();

        Imgproc.cvtColor(processedImg, processedImg, Imgproc.COLOR_BGR2Lab);
        Core.split(processedImg, lab);

        Core.subtract(lab.get(0), lab.get(1), subA);
        Core.subtract(lab.get(0), lab.get(2), subB);

        Core.add(subA, subB, processedImg);
    }


    /* Method scanImage is the only public method of the Scan class.
     *  This method will be called to scan the image provided at time of
     *  constructing Scan class' object.
     * 
     * scanImage method uses switch-case to execute required methods in correct order
     *  to implemnt desired mode of scanning.
     * 
     * This method takes enum type as argument.
     * 
     * whitePointSelect() and blackPointSelect() methods are designed to process "processedImg"
     *  hence "inputImg" is copied to "processedImg" in RMODE and SMODE.
     *  highPassFilter() outputs filtered image as "filtered" hence we need to copy "filtered"
     *  to "processedImg" so that whitePointSelect() can further process it.
     */
    public Mat scanImage(ScanMode mode) {
        
        switch (mode) {
            case GCMODE:
                this.highPassFilter();
                processedImg = filtered.clone();
                // Fix white point value at 127 for GCMODE
                this.whitePoint = 127;
                this.whitePointSelect();
                this.blackPointSelect();
                break;

            case RMODE:
                processedImg = inputImg.clone();
                this.blackPointSelect();
                this.whitePointSelect();
                break;

            case SMODE:
                processedImg = inputImg.clone();
                this.blackPointSelect();
                this.whitePointSelect();
                this.blackAndWhite();
                break;
        
            default:
                System.out.println("Error: Incorrect ScanMode supplied. Expected input: GCSCAN/RSCAN/SCAN");
                break;
        }


    return processedImg;
    }
}
